ddtrace/contrib/internal/langchain/patch.py

```diff
@@ -975,6 +976,165 @@ def traced_similarity_search(langchain, pin, func, instance, args, kwargs):
+    llm_provider = instance._llm_type
+
+    def _on_span_started(span: Span):
+        chat_messages = get_argument_value(args, kwargs, 0, "input")
+        if not isinstance(chat_messages, list):
+            chat_messages = [chat_messages]
+        for message_idx, message in enumerate(chat_messages):
+            if integration.is_pc_sampled_span(span):
+                if isinstance(message, dict):
+                    span.set_tag_str(
+                        "langchain.request.messages.%d.content" % (message_idx),
+                        integration.trunc(str(message.get("content", ""))),
+                    )
+                    span.set_tag_str(
+                        "langchain.request.messages.%d.role" % (message_idx),
+                        str(message.get("role", "")),
+                    )
+                elif isinstance(message, langchain_core.prompt_values.PromptValue):
+                    for langchain_message_idx, langchain_message in enumerate(message.messages):
+                        span.set_tag_str(
+                            "langchain.request.messages.%d.%d.content" % (message_idx, langchain_message_idx),
+                            integration.trunc(str(langchain_message.content)),
+                        )
+                        span.set_tag_str(
+                            "langchain.request.messages.%d.%d.role" % (message_idx, langchain_message_idx),
+                            str(langchain_message.__class__.__name__),
+                        )
+                elif isinstance(message, langchain_core.messages.BaseMessage):
+                    span.set_tag_str(
+                        "langchain.request.messages.%d.content" % (message_idx), integration.trunc(str(message.content))
+                    )
+                    span.set_tag_str(
+                        "langchain.request.messages.%d.role" % (message_idx), str(message.__class__.__name__)
+                    )
+                else:
+                    span.set_tag_str(
+                        "langchain.request.messages.%d.content" % (message_idx), integration.trunc(message)
+                    )
+
+        for param, val in getattr(instance, "_identifying_params", {}).items():
+            if isinstance(val, dict):
+                for k, v in val.items():
+                    span.set_tag_str("langchain.request.%s.parameters.%s.%s" % (llm_provider, param, k), str(v))
+            else:
+                span.set_tag_str("langchain.request.%s.parameters.%s" % (llm_provider, param), str(val))
+
+    def _on_span_finished(span: Span, streamed_chunks, error: Optional[bool] = None):
+        if not error and integration.is_pc_sampled_span(span):
+            content = "".join([str(chunk.content) for chunk in streamed_chunks])
+            span.set_tag_str("langchain.response.content", integration.trunc(content))
+
+            usage = getattr(streamed_chunks[-1], "usage_metadata", None)
```

Comment:
Is usage metadata always guaranteed to be the last chunk?

Reply:
I believe so (if it's included), but I'll try and look to see if there's evidence it could show up in a different chunk

Reply:
this might be openai-specific. since we already have a check in the non-streamed chat model patching to check `_is_openai_chat_instance`, i'll add that here as well, and in that case I believe it is guaranteed to be on the last chunk